### PR TITLE
Compressed

### PR DESCRIPTION
**Pull request by jannikluhn**
### What was wrong?

Missing support for compressed public keys (#55).

### How was it fixed?

- Add `to/from_compressed_bytes` methods to `PublicKey` class.
- Add implementations for the coincurve and native backend.
- Add some tests.

I'm not an expert in cryptography, so please review carefully, especially the native backend code.

#### Cute Animal Picture

Cute animal picture

### CODE REVIEW

File: eth_keys/backends/base.py
Comment (pipermerriam) [inline eth_keys/backends/base.py:33]: Note to self that we should convert these to be `ABC,abstractmethod` based sometime.

File: eth_keys/validation.py
Comment (pipermerriam) [inline eth_keys/validation.py]: Nice to include the actual value length here as well so it's clear what the discrepancy is.

**Review by pipermerriam — COMMENTED** [review]
Looks like there is one more easy test that could be written to randomly generate a private key and validate that `decompress(compress(original)) == original` is always true (using hypothesis)

Also needs to update the documentation with these new APIs.

**Comment by jannikluhn** [comment]
Added the suggested hypothesis test, improved the validation error message, and updated the readme.

Also, when looking at the readme I realized that the key compression API is duplicated: Once in PublicKey (where one would naturally look for it) and once in `KeyAPI` (where it doesn't make too much sense IMO). I therefore removed it from `KeyAPI` and udpated the tests accordingly.

File: eth_keys/backends/native/ecdsa.py
Comment (carver) [inline eth_keys/backends/native/ecdsa.py:69]: lol, padding the compressed key is a strange spec. I was wondering why the compressed key is always the same length...

File: tests/backends/test_native_backend_against_coincurve.py
Comment (carver) [inline tests/backends/test_native_backend_against_coincurve.py]: What's the point of reversing direction on this? Looks like copy-pasta, to me.

**Review by carver — APPROVED** [review]
Will wait a bit to see if Piper has more to add. Looks like one of the tests can be simplified.

File: tests/backends/test_native_backend_against_coincurve.py
Comment (jannikluhn) [reply tests/backends/test_native_backend_against_coincurve.py]: Whoops, you're right, there's no point. Fixed.

**Comment by carver** [comment]
Anything else to add today @pipermerriam ? If not, I'll merge by EOD.

File: eth_keys/backends/coincurve.py
Comment (pipermerriam) [inline eth_keys/backends/coincurve.py:88]: This allows for an arbitrarily long second value were the provided `uncompressed_public_key_bytes` to be longer than `64` bytes.  Am I missing something?  Seems like this would cause weird/hard-to-debug downstream errors in some cases.

File: eth_keys/backends/native/ecdsa.py
Comment (pipermerriam) [inline eth_keys/backends/native/ecdsa.py:80]: I've been on a big `memoryview` kick recently so feel free to ignore me.  This would be slightly faster using a `memoryview` to prevent the copy.  This is only safe if the thing you're passing it to doesn't try to mutate it which in this case I believe that condition is satisfied.

**Review by pipermerriam — APPROVED** [review]
looks good.

The hypothesis tests are probably only running a small number of runs IIRC so you might do a couple of local runs or repeat the hypothesis test runs on a loop for a while to stress test them a bit and see if you find any edge cases that demonstrate the two implementations diverging before you click merge.

File: eth_keys/backends/coincurve.py
Comment (jannikluhn) [reply eth_keys/backends/coincurve.py:88]: This is not part of the public API, all inputs should pass through `datatypes.PublicKey`. I still added validation now, just to be safe.

File: eth_keys/backends/native/ecdsa.py
Comment (jannikluhn) [reply eth_keys/backends/native/ecdsa.py:80]: I did a quick benchmark and didn't see any improvement. I think the reason is that it would save one copy when slicing, but it adds one when creating the `memoryview`. But I will keep it in mind in the future (could be very useful for pyrlp/ssz).

**Comment by jannikluhn** [comment]
I don't have write access here, can someone merge please?

**Comment by carver** [comment]
Would you like a release after #58 is merged?

**Comment by jannikluhn** [comment]
Thanks!

> Would you like a release after #58 is merged?

Yep, that would be nice.

**Pull request merged**
